clang/lib/Driver/ToolChains/HLSL.cpp

```diff
@@ -0,0 +1,160 @@
+using namespace clang;
+using namespace llvm::opt;
+using namespace llvm;
+
+namespace {
+
+const unsigned OfflineLibMinor = 0xF;
+const unsigned MaxShaderModel6Minor = 7;
+
+bool isLegalVersion(VersionTuple Version, unsigned Major, unsigned MinMinor,
+                    unsigned MaxMinor) {
+  VersionTuple Min(Major, MinMinor);
+  VersionTuple Max(Major, MaxMinor);
+  return Min <= Version && Version <= Max;
+}
+
+bool isLegalShaderModel(Triple &T) {
+  if (T.getOS() != Triple::OSType::ShaderModel)
+    return false;
+
+  auto Version = T.getOSVersion();
+  if (Version.getBuild())
+    return false;
+  if (Version.getSubminor())
+    return false;
+  if (!Version.getMinor())
+    return false;
+
+  auto Kind = T.getEnvironment();
+
+  switch (Kind) {
+  default:
+    return false;
+  case Triple::EnvironmentType::Vertex:
+  case Triple::EnvironmentType::Hull:
+  case Triple::EnvironmentType::Domain:
+  case Triple::EnvironmentType::Geometry:
+  case Triple::EnvironmentType::Pixel:
+  case Triple::EnvironmentType::Compute: {
+    if (isLegalVersion(Version, 4, 0, 1))
+      return true;
+    if (isLegalVersion(Version, 5, 0, 1))
+      return true;
+
+    if (isLegalVersion(Version, 6, 0, MaxShaderModel6Minor))
+      return true;
+  } break;
+  case Triple::EnvironmentType::Library: {
+    VersionTuple SM6x(6, OfflineLibMinor);
+    if (Version == SM6x)
+      return true;
+    if (isLegalVersion(Version, 6, 3, MaxShaderModel6Minor))
+      return true;
+  } break;
+  case Triple::EnvironmentType::Amplification:
+  case Triple::EnvironmentType::Mesh: {
+    if (isLegalVersion(Version, 6, 5, MaxShaderModel6Minor))
+      return true;
+  } break;
+  }
+  return false;
+}
+
+std::string tryParseProfile(StringRef Profile) {
+  // [ps|vs|gs|hs|ds|cs|ms|as]_[major]_[minor]
+  SmallVector<StringRef, 3> Parts;
+  Profile.split(Parts, "_");
+  if (Parts.size() != 3)
+    return "";
+
+  Triple::EnvironmentType Kind =
+      StringSwitch<Triple::EnvironmentType>(Parts[0])
+          .Case("ps", Triple::EnvironmentType::Pixel)
+          .Case("vs", Triple::EnvironmentType::Vertex)
+          .Case("gs", Triple::EnvironmentType::Geometry)
+          .Case("hs", Triple::EnvironmentType::Hull)
+          .Case("ds", Triple::EnvironmentType::Domain)
+          .Case("cs", Triple::EnvironmentType::Compute)
+          .Case("lib", Triple::EnvironmentType::Library)
+          .Case("ms", Triple::EnvironmentType::Mesh)
+          .Case("as", Triple::EnvironmentType::Amplification)
+          .Default(Triple::EnvironmentType::UnknownEnvironment);
+  if (Kind == Triple::EnvironmentType::UnknownEnvironment)
+    return "";
+
+  unsigned Major = StringSwitch<unsigned>(Parts[1])
+                       .Case("4", 4)
+                       .Case("5", 5)
+                       .Case("6", 6)
+                       .Default(0);
+  if (Major == 0)
+    return "";
+
+  const unsigned InvalidMinor = -1;
+  unsigned Minor = StringSwitch<unsigned>(Parts[2])
```

Comment:
I don’t think you need string switches for this at all. You can just convert the strings to any integer. That way this code doesn’t need to be updated in the future, we only need to update the validation function.